lib/phoenix/endpoint.ex

```diff
@@ -533,6 +533,35 @@ defmodule Phoenix.Endpoint do
     end
   end
 
+  @doc """
+  Instruments the given function using the instrumentation provided by
+  `endpoint`.
+
+  Usually, users should prefer to instrument events using the `instrument/3`
+  macro defined in every Phoenix endpoint. This macro should only be used for
+  cases when the endpoint is dynamic and not known at compile time:
+
+  ## Examples
+
+      endpoint = MyApp.Endpoint
+      Phoenix.Endpoint.instrument endpoint, :render_view, fn -> ... end
+
+  """
+  defmacro instrument(endpoint_or_conn, event, runtime \\ nil, fun) do
+    compile = Phoenix.Endpoint.Instrument.strip_caller(__CALLER__) |> Macro.escape()
+
+    quote do
+      case unquote(endpoint_or_conn) do
+        %Plug.Conn{private: %{phoenix_endpoint: endpoint}} ->
```

Comment:
At least, because the socket uses the same structure, we can just match on `%{__struct__: struct, private: ...} when struct in [Plug.Conn, Phoenix...]`